File: android/app/src/main/java/com/navi/presentation/viewmodels/MapViewModel.kt
package com.navi.presentation.viewmodels

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.google.android.gms.maps.model.LatLng
import com.navi.data.services.*
import com.navi.domain.models.Place
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class MapViewModel @Inject constructor(
    private val mapboxManager: MapboxManager,
    private val locationManager: LocationManager,
    private val navigationManager: NavigationManager
) : ViewModel() {
    
    // State flows
    private val _userLocation = MutableStateFlow<LatLng?>(null)
    val userLocation: StateFlow<LatLng?> = _userLocation.asStateFlow()
    
    private val _currentRoute = MutableStateFlow<Route?>(null)
    val currentRoute: StateFlow<Route?> = _currentRoute.asStateFlow()
    
    private val _isNavigating = MutableStateFlow(false)
    val isNavigating: StateFlow<Boolean> = _isNavigating.asStateFlow()
    
    private val _searchResults = MutableStateFlow<List<Place>>(emptyList())
    val searchResults: StateFlow<List<Place>> = _searchResults.asStateFlow()
    
    private val _selectedPlace = MutableStateFlow<Place?>(null)
    val selectedPlace: StateFlow<Place?> = _selectedPlace.asStateFlow()
    
    private val _showRoutePreview = MutableStateFlow(false)
    val showRoutePreview: StateFlow<Boolean> = _showRoutePreview.asStateFlow()
    
    private val _mapStyle = MutableStateFlow(MapStyle.STREETS)
    val mapStyle: StateFlow<MapStyle> = _mapStyle.asStateFlow()
    
    private val _showTraffic = MutableStateFlow(false)
    val showTraffic: StateFlow<Boolean> = _showTraffic.asStateFlow()
    
    private val _error = MutableStateFlow<String?>(null)
    val error: StateFlow<String?> = _error.asStateFlow()
    
    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()
    
    // Navigation state
    private val _currentInstruction = MutableStateFlow("")
    val currentInstruction: StateFlow<String> = _currentInstruction.asStateFlow()
    
    private val _remainingDistance = MutableStateFlow("")
    val remainingDistance: StateFlow<String> = _remainingDistance.asStateFlow()
    
    private val _remainingTime = MutableStateFlow("")
    val remainingTime: StateFlow<String> = _remainingTime.asStateFlow()
    
    private val _eta = MutableStateFlow("")
    val eta: StateFlow<String> = _eta.asStateFlow()
    
    init {
        setupBindings()
    }
    
    // MARK: - Setup
    
    private fun setupBindings() {
        // Bind location updates
        viewModelScope.launch {
            locationManager.userLocation.collect { location ->
                _userLocation.value = location
            }
        }
        
        // Bind navigation state
        viewModelScope.launch {
            navigationManager.isNavigating.collect { isNav ->
                _isNavigating.value = isNav
            }
        }
        
        viewModelScope.launch {
            navigationManager.currentRoute.collect { route ->
                _currentRoute.value = route
            }
        }
        
        viewModelScope.launch {
            navigationManager.currentInstruction.collect { instruction ->
                _currentInstruction.value = instruction
            }
        }
        
        // Bind formatted navigation info
        viewModelScope.launch {
            navigationManager.remainingDistance.collect {
                _remainingDistance.value = navigationManager.getFormattedRemainingDistance()
            }
        }
        
        viewModelScope.launch {
            navigationManager.remainingDuration.collect {
                _remainingTime.value = navigationManager.getFormattedRemainingTime()
            }
        }
        
        viewModelScope.launch {
            navigationManager.estimatedArrivalTime.collect {
                _eta.value = navigationManager.getFormattedETA()
            }
        }
    }
    
    // MARK: - Location Actions
    
    fun startLocationTracking() {
        locationManager.startTracking()
    }
    
    fun centerOnUserLocation(mapView: com.mapbox.maps.MapView) {
        val location = _userLocation.value ?: return
        mapboxManager.centerOnLocation(mapView, location.latitude, location.longitude)
    }
    
    // MARK: - Route Actions
    
    fun calculateRoute(destination: LatLng) {
        val origin = _userLocation.value
        if (origin == null) {
            _error.value = "Unable to get your current location"
            return
        }
        
        _isLoading.value = true
        
        viewModelScope.launch {
            val result = mapboxManager.calculateRoute(
                origin.latitude,
                origin.longitude,
                destination.latitude,
                destination.longitude
            )
            
            _isLoading.value = false
            
            result.onSuccess { route ->
                _currentRoute.value = route
                _showRoutePreview.value = true
            }
            
            result.onFailure { error ->
                _error.value = error.message
            }
        }
    }
    
    fun calculateRouteToPlace(place: Place) {
        val destination = LatLng(place.latitude, place.longitude)
        calculateRoute(destination)
    }
    
    fun startNavigation() {
        val route = _currentRoute.value ?: return
        navigationManager.startNavigation(route)
        _showRoutePreview.value = false
    }
    
    fun stopNavigation() {
        navigationManager.stopNavigation()
        mapboxManager.clearRoute()
        _currentRoute.value = null
    }
    
    fun pauseNavigation() {
        navigationManager.pauseNavigation()
    }
    
    fun resumeNavigation() {
        navigationManager.resumeNavigation()
    }
    
    // MARK: - Map Style
    
    fun changeMapStyle(mapView: com.mapbox.maps.MapView, style: MapStyle) {
        _mapStyle.value = style
        mapboxManager.setMapStyle(mapView, style)
    }
    
    fun toggleTraffic() {
        _showTraffic.value = !_showTraffic.value
        // Traffic toggle implementation would go here
    }
    
    // MARK: - Place Selection
    
    fun selectPlace(place: Place) {
        _selectedPlace.value = place
        
        // Add marker on map
        mapboxManager.clearMarkers()
        mapboxManager.addMarker(place.latitude, place.longitude, place.name)
    }
    
    fun deselectPlace() {
        _selectedPlace.value = null
        mapboxManager.clearMarkers()
    }
    
    // MARK: - Voice Guidance
    
    fun toggleVoiceGuidance() {
        navigationManager.toggleVoiceGuidance()
    }
    
    fun setVoiceVolume(volume: Float) {
        navigationManager.setVoiceVolume(volume)
    }
    
    // MARK: - Cleanup
    
    override fun onCleared() {
        super.onCleared()
        locationManager.stopTracking()
    }
}

File: ios/Navi/ViewModels/MapViewModel.swift
import Foundation
import CoreLocation
import Combine
import MapboxMaps

/// Map ViewModel - Manages map screen state and user interactions
class MapViewModel: ObservableObject {
    // MARK: - Published Properties
    @Published var userLocation: CLLocationCoordinate2D?
    @Published var currentRoute: Route?
    @Published var isNavigating = false
    @Published var searchResults: [Place] = []
    @Published var selectedPlace: Place?
    @Published var showRoutePreview = false
    @Published var mapStyle: MapStyle = .streets
    @Published var showTraffic = false
    @Published var error: String?
    @Published var isLoading = false
    
    // Navigation state
    @Published var currentInstruction: String = ""
    @Published var remainingDistance: String = ""
    @Published var remainingTime: String = ""
    @Published var eta: String = ""
    
    // MARK: - Private Properties
    private var cancellables = Set<AnyCancellable>()
    private let mapboxManager = MapboxManager.shared
    private let locationManager = LocationManager.shared
    private let navigationManager = NavigationManager.shared
    
    // MARK: - Initialization
    init() {
        setupBindings()
    }
    
    // MARK: - Setup
    
    private func setupBindings() {
        // Bind location updates
        locationManager.$userLocation
            .assign(to: &$userLocation)
        
        // Bind navigation state
        navigationManager.$isNavigating
            .assign(to: &$isNavigating)
        
        navigationManager.$currentRoute
            .assign(to: &$currentRoute)
        
        navigationManager.$currentInstruction
            .assign(to: &$currentInstruction)
        
        // Bind formatted navigation info
        navigationManager.$remainingDistance
            .map { _ in navigationManager.getFormattedRemainingDistance() }
            .assign(to: &$remainingDistance)
        
        navigationManager.$remainingDuration
            .map { _ in navigationManager.getFormattedRemainingTime() }
            .assign(to: &$remainingTime)
        
        navigationManager.$estimatedArrivalTime
            .map { _ in navigationManager.getFormattedETA() }
            .assign(to: &$eta)
    }
    
    // MARK: - Location Actions
    
    func requestLocationPermission() {
        locationManager.requestPermissions()
    }
    
    func startLocationTracking() {
        locationManager.startTracking()
    }
    
    func centerOnUserLocation() {
        guard let location = userLocation else { return }
        mapboxManager.centerOnUserLocation(location)
    }
    
    // MARK: - Route Actions
    
    func calculateRoute(to destination: CLLocationCoordinate2D) {
        guard let origin = userLocation else {
            error = "Unable to get your current location"
            return
        }
        
        isLoading = true
        
        mapboxManager.calculateRoute(from: origin, to: destination) { [weak self] result in
            self?.isLoading = false
            
            switch result {
            case .success(let route):
                self?.currentRoute = route
                self?.showRoutePreview = true
                self?.mapboxManager.displayRoute(route)
            case .failure(let error):
                self?.error = error.localizedDescription
            }
        }
    }
    
    func calculateRouteToPlace(_ place: Place) {
        let destination = CLLocationCoordinate2D(
            latitude: place.latitude,
            longitude: place.longitude
        )
        calculateRoute(to: destination)
    }
    
    func startNavigation() {
        guard let route = currentRoute else { return }
        
        navigationManager.startNavigation(route: route)
        showRoutePreview = false
    }
    
    func stopNavigation() {
        navigationManager.stopNavigation()
        mapboxManager.clearRoute()
        currentRoute = nil
    }
    
    func pauseNavigation() {
        navigationManager.pauseNavigation()
    }
    
    func resumeNavigation() {
        navigationManager.resumeNavigation()
    }
    
    // MARK: - Map Style
    
    func changeMapStyle(_ style: MapStyle) {
        mapStyle = style
        mapboxManager.setMapStyle(style)
    }
    
    func toggleTraffic() {
        showTraffic.toggle()
        mapboxManager.toggleTraffic(enabled: showTraffic)
    }
    
    // MARK: - Place Selection
    
    func selectPlace(_ place: Place) {
        selectedPlace = place
        
        // Add marker on map
        let coordinate = CLLocationCoordinate2D(
            latitude: place.latitude,
            longitude: place.longitude
        )
        mapboxManager.clearMarkers()
        mapboxManager.addMarker(at: coordinate, title: place.name)
        
        // Center on place
        mapboxManager.centerOnUserLocation(coordinate, zoom: 14)
    }
    
    func deselectPlace() {
        selectedPlace = nil
        mapboxManager.clearMarkers()
    }
    
    // MARK: - Voice Guidance
    
    func toggleVoiceGuidance() {
        navigationManager.toggleVoiceGuidance()
    }
    
    func setVoiceVolume(_ volume: Float) {
        navigationManager.setVoiceVolume(volume)
    }
}
